app/services/atmosphere/billing_service.rb

```diff
@@ -20,45 +20,57 @@ def self.bill_all_appliances
+    # apply_prepayment determines whether the deployment should be prepaid
+    # (false when shutting down appliance or deployment)
     def self.bill_appliance(appl, billing_time, message, apply_prepayment = true)
-      if appl.prepaid_until.blank?
-        # This should not happen - it means that we do not know when this appliance will run out of funds. Therefore we cannot meaningfully bill it again and must return an error.
-        Rails.logger.error("Unable to determine current payment validity date for appliance #{appl.id}. Skipping.")
-        BillingLog.new(timestamp: Time.now.utc, user: appl.appliance_set.user, appliance: appl.id.to_s+'---'+(appl.name.blank? ? 'unnamed_appliance' : appl.name), fund: appl.fund.name, message: "Unable to determine current payment validity date for appliance.", actor: "bill_appliance", amount_billed: 0).save
-        appl.billing_state = "error"
-        appl.save
-        raise Atmosphere::BillingException.new(message: "Unable to determine current payment validity date for appliance #{appl.id}")
+      appl.deployments.each do |dep|
+        bill_deployment(dep, billing_time, message, apply_prepayment)
+      end
+    end
+
+
+    def self.bill_deployment(dep, billing_time, message, apply_prepayment = true)
+      appl = dep.appliance # Shorthand
+      if dep.prepaid_until.blank?
+        # This should not happen - it means that we do not know when this deployment will run out of funds.
+        # Therefore we cannot meaningfully bill it again and must return an error.
+        Rails.logger.error("Unable to determine current payment validity date for deployment #{dep.id} belonging to appliance #{appl.id}. Skipping.")
+        BillingLog.new(timestamp: Time.now.utc, user: appl.appliance_set.user, appliance: appl.id.to_s+'---'+(appl.name.blank? ? 'unnamed_appliance' : appl.name), fund: appl.fund.name, message: "Unable to determine current payment validity date for deployment.", actor: "bill_appliance", amount_billed: 0).save
+        dep.billing_state = "error"
+        dep.save
+        raise Atmosphere::BillingException.new(message: "Unable to determine current payment validity date for deployment #{dep.id} belonging to appliance #{appl.id}")
       else
-        # Figure out how long this appliance can continue to run without being billed again.
-        billing_interval = (billing_time - appl.prepaid_until) # This will return time in seconds
+        # Figure out how long this deployment can continue to run without being billed again.
+        billing_interval = (billing_time - dep.prepaid_until) # This will return time in seconds
         if billing_interval < 0
           # The appliance is still prepaid - nothing to be done.
-          Rails.logger.debug("Appliance #{appl.id} is prepaid until #{appl.prepaid_until}. Nothing to be done.")
+          Rails.logger.debug("Deployment #{dep.id} belonging to appliance #{appl.id} is prepaid until #{dep.prepaid_until}. Nothing to be done.")
```

Comment:
For debug log with string concatenation please use following structure:

```
Rails.logger.debug { "this is #{my} debug message" }
```

of (no newlines thanks to `squish`)

```ruby
Rails.logger.debug do
  <<-LOG.squish
  My long
  debug message
  with #{variables}
  LOG
end
```

the difference is that example presented above will evaluate lambda only when debug is turned on. In original code string in interpolated every time, even if debug is turned off.